Framework/Algorithms/src/LorentzCorrection.cpp

```diff
@@ -45,58 +68,99 @@ const std::string LorentzCorrection::name() const {
+}
+
+std::map<std::string, std::string> LorentzCorrection::validateInputs() {
+  std::map<std::string, std::string> result;
+
+  const auto processingType = this->getPropertyValue(PropertyNames::TYPE);
+  // check units if the SCD option is selected
+  if (processingType == TOF_SCD) {
+    MatrixWorkspace_const_sptr wksp =
+        this->getProperty(PropertyNames::INPUT_WKSP);
+    // code is a variant of private method from WorkspaceUnitValidator
+    const auto unit = wksp->getAxis(0)->unit();
+    if ((!unit) || (unit->unitID().compare("Wavelength"))) {
+      result[PropertyNames::INPUT_WKSP] =
+          "The workspace must have units of Wavelength";
+    }
+  }
+
+  return result;
 }
 
 /** Execute the algorithm.
  */
 void LorentzCorrection::exec() {
-  MatrixWorkspace_sptr inWS = this->getProperty("InputWorkspace");
+  MatrixWorkspace_sptr inWS = this->getProperty(PropertyNames::INPUT_WKSP);
 
+  // clone the workspace - does nothing for inplace operation
   auto cloneAlg = this->createChildAlgorithm("CloneWorkspace", 0, 0.1);
   cloneAlg->initialize();
   cloneAlg->setProperty("InputWorkspace", inWS);
+  cloneAlg->setPropertyValue(
+      "OutputWorkspace", this->getPropertyValue(PropertyNames::OUTPUT_WKSP));
   cloneAlg->execute();
   Workspace_sptr temp = cloneAlg->getProperty("OutputWorkspace");
   MatrixWorkspace_sptr outWS =
       boost::dynamic_pointer_cast<MatrixWorkspace>(temp);
 
-  const auto numHistos = inWS->getNumberHistograms();
-  const auto &spectrumInfo = inWS->spectrumInfo();
-  Progress prog(this, 0.0, 1.0, numHistos);
+  Progress prog(this, 0.1, 1.0, inWS->getNumberHistograms());
+
+  const auto processingType = this->getPropertyValue(PropertyNames::TYPE);
+  if (processingType == TOF_SCD)
+    processTOF_SCD(outWS, prog);
+  else if (processingType == TOF_PD)
+    processTOF_PD(outWS, prog);
+  else {
+    std::stringstream msg;
+    msg << "Do not understand know how to process Type=\"" << processingType
+        << "\" - developer forgot to fill in if/else tree";
+    throw std::runtime_error(msg.str());
+  }
+
+  this->setProperty(PropertyNames::OUTPUT_WKSP, outWS);
+}
 
-  PARALLEL_FOR_IF(Kernel::threadSafe(*inWS))
-  for (int64_t i = 0; i < int64_t(numHistos); ++i) {
+void LorentzCorrection::processTOF_SCD(MatrixWorkspace_sptr &wksp,
+                                       Progress &prog) {
+  const int64_t numHistos = static_cast<int64_t>(wksp->getNumberHistograms());
+  const auto &spectrumInfo = wksp->spectrumInfo();
+
+  PARALLEL_FOR_IF(Kernel::threadSafe(*wksp))
+  for (int64_t i = 0; i < numHistos; ++i) {
     PARALLEL_START_INTERUPT_REGION
 
     if (!spectrumInfo.hasDetectors(i))
       continue;
 
-    const double twoTheta =
-        spectrumInfo.isMonitor(i) ? 0.0 : spectrumInfo.twoTheta(i);
-    const double sinTheta = std::sin(twoTheta / 2);
-    double sinThetaSq = sinTheta * sinTheta;
+    const double sinTheta_v = sinTheta(spectrumInfo, i);
```

Comment:
You need to handle the monitor case. A monitor at 180 degrees will have a sinTheta_v=1. I suggest that you just zero the spectra.